### ISSUE TITLE
P2SH Litecoin Address considered as Invalid 

### ISSUE THREAD

**Issue by goodProjects** [Bit-Wasp/bitcoin-php#677]
```
   Bitcoin::setNetwork(NetworkFactory::litecoin());
   $address = new AddressCreator();
        try {
            $script = $address->fromString('34QnEPVFnBboiUgsBayTo9BLesjKWKo84T')->getScriptPubKey()->getBuffer();
        }catch (\Exception $e){
            throw new InvalidAddress();
        }
```

is that an expected behavior?

**Comment by afk11** [comment]
I recently upgraded to their new prefix, kinda figured this would happen.

You could override the readBase58Address function from AddressCreator in a litecoin specific one - LitecoinAddressCreator. This has the downside that if you do get a ScriptHashAddress for LTC and encode it again, it would use the constants on the network and not your one..

Or, you could just implement OldLitecoin with the prefix?

```php
    protected function readBase58Address(string $strAddress, NetworkInterface $network)
    {
        try {
            $data = Base58::decodeCheck($strAddress);
            $prefixByte = $data->slice(0, 1)->getHex();
            if ($network instanceof Litecoin && $prefixByte == '05') {
                return new ScriptHashAddress($data->slice(1));
            } else if ($network instanceof LitecoinTestnet) {
                // .........
            } else if ($prefixByte === $network->getP2shByte()) {
                return new ScriptHashAddress($data->slice(1));
            } else if ($prefixByte === $network->getAddressByte()) {
                return new PayToPubKeyHashAddress($data->slice(1));
            }
        } catch (\Exception $e) {
            // Just return null
        }
        return null;
    }
```

**Comment by goodProjects** [comment]
Thank you so much for your help!

**Comment by goodProjects** [comment]
I've made the changes it's working, But why don't we apply that to this library so we have those addresses support?

```
  protected function readBase58Address($strAddress, NetworkInterface $network)
    {
        try {
            $data = Base58::decodeCheck($strAddress);
            $prefixByte = $data->slice(0, 1)->getHex();

            if ($network instanceof Litecoin && $prefixByte === '05') {
                return new ScriptHashAddress($data->slice(1));
            } else if ($network instanceof LitecoinTestnet  && $prefixByte === '6f') {
                return new ScriptHashAddress($data->slice(1));
            }else if ($prefixByte === $network->getP2shByte()) {
                return new ScriptHashAddress($data->slice(1));
            } else if ($prefixByte === $network->getAddressByte()) {
                return new PayToPubKeyHashAddress($data->slice(1));
            }

        } catch (\Exception $e) {
            // Just return null
        }

        return null;
    }
```

**Comment by afk11** [comment]
Then you couldn't shut it off :) Generally, I think it's best to keep network specific code hidden behind a base type, so people can bring their own implementations. 

I learned this the hard way when I realized I wouldn't be able to support all the hardforks of bitcoin (not that I want to), and that by sticking `new BitcoinCashChecker` in the middle of Signer, I had missed my chance to allow every other network, too! 

This was what lead to CheckerCreator as a base type - now people can write an implementation and Creator and start using it with the Signer class! 

So generally I'm trending towards keeping the library bitcoin specific, but accept abstractions where reasonable to allow other networks!
